### PR TITLE
ENH: minimize_ipopt: add input validation for `bounds` and support for new-style constraints

### PR DESCRIPTION
**Pull request by mdhaber**
Followup to gh-206. Adds input validation for the `bounds` parameter of `minimize_ipopt` and support for `scipy.optimize.Bounds` objects.

### CODE REVIEW

File: cyipopt/scipy_interface.py [hunk 1]
```diff
@@ -486,13 +486,11 @@ def minimize_ipopt(fun,
     (fun, x0, args, kwargs, method, jac, hess, hessp,
      bounds, constraints, tol, callback, options) = res
 
-    _x0 = np.atleast_1d(x0)
```
Comment (mdhaber) [inline cyipopt/scipy_interface.py:489]: Conversion to 1d is now done in `_minimize_ipopt_iv`.

Previously, the variable was renamed to `_x0` so that at the end, if `x0` had been a scalar, the `x` attribute of the result object would be a scalar.

In SciPy, I am very careful about shapes, and I have corrected many shape bugs related to this sort of thing in SciPy code. However, I'm not sure if this is needed here. 

The documentation (which now mirrors that of [`scipy.optimize.minimize`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html)) states that `x0` should be an array-like of shape `(n,)`, where `n` is the number of independent variables. This function does not handle arrays of arbitrary dimensionality, and there is a separate `minimize_scalar` function for functions of a single variable. Also, the `x` attribute of the object returned by `scipy.optimize.minimize` returns 1d array even when x0 is a scalar:

```python3
from scipy.optimize import minimize

def f(x):
    return x**2

minimize(f, 1).x  # [-7.450e-09]
```

All these things suggest to me that the output should always be a 1d array (even if the function is flexible about accepting scalar input rather than raising an error)and working with functions that return scalars).

LMK if you'd prefer for this to be left alone.

File: cyipopt/scipy_interface.py [hunk 2]
```diff
@@ -555,20 +550,50 @@ def minimize_ipopt(fun,
...
+        lb, ub = bounds.lb, bounds.ub
+    else:
+        bounds = np.atleast_2d(bounds)
+        if bounds.shape[1] != 2:
+            raise ValueError("`bounds` must specify both lower and upper "
+                             "limits for each decision variable.")
+        lb, ub = bounds.T
+
+    try:
+        lb, ub, x0 = np.broadcast_arrays(lb, ub, x0)
```
Comment (mdhaber) [inline cyipopt/scipy_interface.py:580]: Somewhat related to the comment above: here, I'm making the function even more flexible about input shape. Previously, the function would raise an error if the user passes only a single lower and upper bound but there are many decision variables; this broadcasts the bounds to apply to all the decision variables.

`scipy.optimize.minimize` is more strict than this, but it doesn't do a great job of validating the input, so you get some imperfect error messages. It's also sort of a mess in this department, so I'm not sure if it's been thought through very carefully. 

LMK if you'd prefer to stay strict and require the number of bounds to equal the number of decision variables rather than following broadcasting rules.

File: cyipopt/scipy_interface.py [hunk 3]
```diff
@@ -555,20 +550,50 @@ def minimize_ipopt(fun,
...
+    except ValueError:
+        raise ValueError("The number of lower bounds, upper bounds, and "
+                         "decision variables must be equal or broadcastable.")
+
+    try:
+        lb = lb.astype(np.float64)
+        ub = ub.astype(np.float64)
```
Comment (mdhaber) [inline cyipopt/scipy_interface.py:587]: Does IPOPT support other dtypes? 
I could try to maintain user floating point dtype here. The reason for this step is that if the user passes `bounds` as a list of lists with `None` to represent no limit / infinity, then `lb` and `ub` get converted to object arrays.

**Review by mdhaber — COMMENTED** [review]
Some self-review to aid the reviewer.

**Comment by mdhaber** [comment]
I think I'd like to remove `Bounds` from this. There is enough here without it, and I want to make sure that the new-style constraints (`LinearConstraint`, `NonlinearConstraint`) can be integrated smoothly before adding `Bounds`.

I could still use feedback on the rest, and I can address it all together.

File: cyipopt/scipy_interface.py [hunk 4]
```diff
@@ -129,7 +129,7 @@ def __init__(self,
 
         if hess is not None:
             self.obj_hess = hess
-        if jac is None:
+        if not jac:
```
Comment (mdhaber) [inline cyipopt/scipy_interface.py:132]: When `jac` is `False`, the code currently treats it as though it were a callable.

File: cyipopt/scipy_interface.py [hunk 5]
```diff
@@ -522,7 +522,9 @@ def minimize_ipopt(fun,
     # Rename some default scipy options
     replace_option(options, b'disp', b'print_level')
     replace_option(options, b'maxiter', b'max_iter')
-    if b'print_level' not in options:
+    if getattr(options, 'print_level', False) is True:
```
Comment (mdhaber) [inline cyipopt/scipy_interface.py:579]: When `disp` is `True`, we get an error because `print_level` needs to be a bool.

File: cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py [hunk 6]
```diff
@@ -31,6 +30,8 @@ class TestSLSQP:
     This example maximizes the function f(x) = 2*x*y + 2*x - x**2 - 2*y**2,
     which has a maximum at x=2, y=1.
     """
+    atol = 1e-7
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py:33]: SLSQP passes these tests without a tolerance, but Ipopt currently needs one.

File: cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py [hunk 7]
```diff
@@ -99,70 +100,70 @@ def fprime_ieqcon2(self, x):
 
     # minimize
     def test_minimize_unbounded_approximated(self):
-        # Minimize, method='SLSQP': unbounded, approximated jacobian.
-        jacs = [None, False, '2-point', '3-point']
+        # Minimize, method=None: unbounded, approximated jacobian.
+        jacs = [None, False]
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py:104]: I suppose if we want the interface to exactly match `minimize`, we need to accept these strings.

File: cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py [hunk 8]
```diff
@@ -197,39 +198,21 @@ def test_minimize_inequality_given(self):
...
-                       method='SLSQP', args=(-1.0,),
+                       method=None, args=(-1.0,),
                        constraints={'type': 'ineq',
                                     'fun': self.f_ieqcon2,
                                     'jac': self.fprime_ieqcon2},
                        options=self.opts)
         assert_(res['success'], res['message'])
         assert_allclose(res.x, [2, 1])
 
-    def test_minimize_bounded_constraint(self):
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py]: Ipopt does not evaluate strictly within the bounds.

File: cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py [hunk 9]
```diff
@@ -239,127 +222,8 @@ def test_minimize_bound_equality_given2(self):
                        options=self.opts)
         assert_(res['success'], res['message'])
         assert_allclose(res.x, [0.8, 0.8], atol=1e-3)
-        assert_(-0.8 <= res.x[0] <= 1)
-        assert_(-1 <= res.x[1] <= 0.8)
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py]: These lines are not really deleted; tolerances are added.

File: cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py [hunk 10]
```diff
@@ -239,127 +222,8 @@ def test_minimize_bound_equality_given2(self):
...
-
-    # fmin_slsqp
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py]: Delete all these `fmin_slsqp` tests; we only want tests of the `minimize` interface.

File: cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py [hunk 11]
```diff
@@ -465,29 +332,21 @@ def f(x):
         cons_ul = [{'type': 'ineq', 'fun': lambda x: 0 - x},
                    {'type': 'ineq', 'fun': lambda x: x + 1}]
 
-        sol = minimize(f, [10], method='slsqp', constraints=cons_u)
-        assert_(sol.success)
-        assert_allclose(sol.x, 0, atol=1e-10)
-
-        sol = minimize(f, [-10], method='slsqp', constraints=cons_l)
-        assert_(sol.success)
-        assert_allclose(sol.x, 2, atol=1e-10)
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py]: The changes below look substantial, but really I only changed `method` to `None` and removed the last two tests because Ipopt thinks they are infeasible.

File: cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py [hunk 12]
```diff
@@ -582,27 +444,8 @@ def test_gh9640(self):
...
             return 1
         x0 = [-1.8869783504471584, -0.640096352696244, -0.8174212253407696]
-        res = minimize(target, x0, method='SLSQP', bounds=bnds, constraints=cons,
+        res = minimize(target, x0, method=None, bounds=bnds, constraints=cons,
                        options={'disp':False, 'maxiter':10000})
 
         # The problem is infeasible, so it cannot succeed
         assert not res.success
-
-    def test_parameters_stay_within_bounds(self):
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_from_scipy.py]: IPOPT doesn't warn, so this test is not needed.

**Review by mdhaber — COMMENTED** [review]
Comments about changes to the tests copied from SciPy.

File: cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py [hunk 13]
```diff
@@ -443,178 +446,97 @@ def hess(x, v):
...
-                        continue
-                    if prob.grad is True and grad in ('3-point', False):
-                        continue
-                    with suppress_warnings() as sup:
-                        sup.filter(UserWarning, "delta_grad == 0.0")
-                        result = minimize(prob.fun, prob.x0,
+            for grad in (prob.grad, False):
+                if prob == list_of_problems[1]:  # MaratosGradInFunc
+                    grad = True
+                for hess in (None,):
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py:468]: We can't use any of the Hessian options now because a) if `hess` is provided, all the constraint Hessians must be provided and b) constraint Hessians can't be transferred from `NonlinearConstraint` to the old constraint format (because the old constraint format in SciPy doesn't ever include constraint Hessians).

It seems possible to get around this, but that would be another PR.

File: cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py [hunk 14]
```diff
@@ -443,178 +446,97 @@ def hess(x, v):
...
-        result2 = minimize(prob.fun, prob.x0,
-                           method='L-BFGS-B',
-                           jac='3-point')
+                          method=None,
+                          jac=prob.grad)
         assert_array_almost_equal(result.x, prob.x_opt, decimal=5)
-        assert_array_almost_equal(result1.x, prob.x_opt, decimal=5)
-        assert_array_almost_equal(result2.x, prob.x_opt, decimal=5)
-
-    def test_hessp(self):
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py]: No support for `hessp`.

File: cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py [hunk 15]
```diff
@@ -443,178 +446,97 @@ def hess(x, v):
...
-                           jac='2-point')
-
-        result2 = minimize(prob.fun, prob.x0,
-                           method='L-BFGS-B',
-                           jac='3-point')
+                          method=None,
+                          jac=prob.grad)
         assert_array_almost_equal(result.x, prob.x_opt, decimal=5)
-        assert_array_almost_equal(result1.x, prob.x_opt, decimal=5)
-        assert_array_almost_equal(result2.x, prob.x_opt, decimal=5)
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py]: Don't need to test L-BFGS-B here.

File: cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py [hunk 16]
```diff
@@ -443,178 +446,97 @@ def hess(x, v):
...
-                          hess=lambda x: 2, callback=callback,
-                          method='trust-constr')
+                          hess=lambda x: 2, method=None)
         assert_(result.get('success'))
-        assert_(result.get('nit', -1) == 1)
+        assert result.get('nit', -1) == 0
 
-        # Also check existence of the 'niter' attribute, for backward
-        # compatibility
-        assert_(result.get('niter', -1) == 1)
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py]: I think this was because `trust-constr` used to have the wrong name for this attribute. In any case, not needed here.

File: cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py [hunk 17]
```diff
@@ -443,178 +446,97 @@ def hess(x, v):
...
-            assert_('nit' in info)
-            assert_('niter' in info)
-
         result = minimize(lambda x: x**2, [0], jac=lambda x: 2*x,
-                          hess=lambda x: 2, callback=callback,
-                          method='trust-constr')
+                          hess=lambda x: 2, method=None)
         assert_(result.get('success'))
-        assert_(result.get('nit', -1) == 1)
+        assert result.get('nit', -1) == 0
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py:535]: I guess IPOPT solves this in 0 iterations?

File: cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py [hunk 18]
```diff
@@ -689,22 +608,20 @@ def assert_inbounds(x):
         assert np.all(x <= bnds.ub)
 
     def obj(x):
-        assert_inbounds(x)
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py]: Ipopt doesn't seem to respect bounds at every iteration.

File: cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py [hunk 19]
```diff
@@ -713,10 +630,14 @@ def nci(x):
 
     ref = minimize(fun=obj, x0=x0, method='slsqp',
                    bounds=bnds, constraints=nlcs)
-    assert_allclose(res.fun, ref.fun)
+    assert ref.success
+    assert res.fun <= ref.fun
```
Comment (mdhaber) [inline cyipopt/tests/unit/test_scipy_ipopt_trust_constr.py:634]: Ipopt finds a much better solution than `slsqp`, apparently.

**Comment by mdhaber** [comment]
This looks like a crazy PR, but I don't think it's actually as bad as it looks.

Almost all the changes to `minimize_ipopt` are in the first commit, so I'd suggest reviewing that first - https://github.com/mechmotum/cyipopt/pull/207/commits/9a01fa627c2fbda97b026b3bea1aef292a44eca1.
I am happy with it, but there are some questions about how strict you want the input validation to be. Currently, it's a bit more flexible than `scipy.optimize.minimize` because it broadcasts compatible `x0`/`bounds` shapes rather than raising an error, but this is what I would have done if writing `minimize` from scratch, and I believe it's compatible with all cases that `minimize` would accept.

The next commit simply adds a test file from SciPy verbatim, so it probably doesn't need to be reviewed from scratch. The commit after that, https://github.com/mechmotum/cyipopt/pull/207/commits/270af6aa0635068ef234f318cd9d3c16d31def40, makes the adjustments needed to make the tests pass; that would be worth reviewing. The tests caught some minor bugs, so there are a few corrections to `minimize_ipopt`.

I think that the only other commit that needs to be reviewed is the last one, https://github.com/mechmotum/cyipopt/pull/207/commits/17e8a328e82b709cdddf14e4f7b5c8ba362ec939.
Again, the one before that just adds a test file from SciPy verbatim, so it just makes the changes needed to make tests pass.

I see that there is a failure on some jobs, but I can't tell what's wrong with that version of the software. The purpose of the test is to make sure that the correct result is produced when the bounds fully specify the solution. Can we xfail the test in those jobs?

**Comment by mdhaber** [comment]
@moorepants just thought I'd send a ping. I know this looks pretty extensive, but I wrote in https://github.com/mechmotum/cyipopt/pull/207#issuecomment-1566406440 why it shouldn't be that bad to review.

After this, I think I'm done with what I had planned to make `minimize_ipopt` a nearly drop-in replacement for `minimize`. There are a few other things that could be done, but I wouldn't add them unless it's requested.

**Comment by moorepants** [comment]
I have not looked at this and don't recall the status. If it is ready for review I could have a look during August. It looks like a number of CI runs have failed though.

**Comment by mdhaber** [comment]
I think it's ready for review because I'd like to know how you'd like to handle the failure produced by `test_equal_all_bounds`. I can't reproduce it locally, even with the same version of SciPy, so I'm not sure what the problem is. Here is a snippet that replicates the test, which I've xfailed in the meantime.

```python3
from scipy.optimize import rosen, Bounds
from cyipopt import minimize_ipopt
bounds = Bounds([4.0, 5.0], [4.0, 5.0])
minimize_ipopt(rosen, [-10, 8], bounds=bounds)
# message: b'Algorithm terminated successfully at a locally optimal point, satisfying the convergence tolerances (can be specified by options).'
# success: True
#  status: 0
#     fun: 12109.0
#       x: [ 4.000e+00  5.000e+00]
#     nit: 0
#    info:     status: 0
#                   x: [ 4.000e+00  5.000e+00]
#                   g: []
#             obj_val: 12109.0
#              mult_g: []
#            mult_x_L: [ 1.761e+04  0.000e+00]
#            mult_x_U: [ 0.000e+00  2.200e+03]
#          status_msg: b'Algorithm terminated successfully at a locally optimal point, satisfying the convergence tolerances (can be specified by options).'
#    nfev: 1
#    njev: 3
```

Can you reproduce the error?

I've pushed a commit that will help me figure out what's going on with the other test.

**Comment by moorepants** [comment]
> The vast majority of the lines are tests copy-pasted from SciPy.

If you are copying lines from SciPy we need to include SciPy's license in this repo. I'm not sure if we discussed this before, but ideally we don't have to have two licenses in this code base. Are the copy paste lines from SciPy necessary? Can you not import from SciPy instead?

**Comment by moorepants** [comment]
> Also, would you be willing to set a minimum supported SciPy version? 

Yes. My general approach has always been to support the SciPy version that shipped with the most recent Ubuntu LTS. It looks like SciPy 1.8.0 was with 22.04 LTS.

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 20]
```diff
@@ -33,9 +33,11 @@ def test_minimize_ipopt_import_error_if_no_scipy():
 @pytest.mark.skipif("scipy" not in sys.modules,
                     reason="Test only valid if Scipy available.")
 def test_minimize_ipopt_input_validation():
+    from scipy import optimize
+
     x0 = 1
     def f(x):
-        return x**2
+        return x @ x
```
Comment (moorepants) [inline cyipopt/tests/unit/test_scipy_optional.py:40]: Is this changing functionality?

```
In [3]: import numpy as np

In [4]: x = np.array([1])

In [5]: x @ x
Out[5]: 1

In [6]: type(x @ x)
Out[6]: numpy.int64

In [7]: type(x**2)
Out[7]: numpy.ndarray
```

**Comment by moorepants** [comment]
One option for the SciPy license is that we do not ship these two test files with the released tarball. We could include the SciPy license in the test directory for those two SciPy (if you download from this repository) but we specifically do not include the files in the release and thus avoid having to add the SciPy license to the release. If possible, I think that is my preference.

File: cyipopt/scipy_interface.py (hunk 3 above)
Comment (moorepants) [reply cyipopt/scipy_interface.py:644]: The types are all set to double for these things:

https://github.com/mechmotum/cyipopt/blob/master/cyipopt/cython/ipopt_wrapper.pyx#L33

File: cyipopt/scipy_interface.py (hunk 2 above)
Comment (moorepants) [reply cyipopt/scipy_interface.py:637]: As long as it works and is tested you can make it flexible.

File: cyipopt/scipy_interface.py (hunk 1 above)
Comment (moorepants) [reply cyipopt/scipy_interface.py:538]: If our function here supported scalar and array objective functions simultaneously, I think we'd need to deprecate the scalar functionality with at least a release.

Comment (mdhaber) [reply cyipopt/scipy_interface.py:538]: This would not change what types of objective functions are supported. It makes the function more consistent with the documentation, which states:

image

In SciPy, the `x` attribute of the result object is _always_ at least 1D. However, currently in `minimize_ipopt`, the `x` attribute of the result object may be a scalar. The proposed change would resolve this inconsistency and make `minimize_ipopt` behave as advertised - like `scipy.optimize.minimize`.

File: cyipopt/scipy_interface.py (hunk 3 above)
Comment (mdhaber) [reply cyipopt/scipy_interface.py:644]: OK, then this line should be fine.

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 20 above)
Comment (mdhaber) [reply cyipopt/tests/unit/test_scipy_optional.py:40]: I think I changed this line just to generalize it to be a generic multivariate quadratic bowl rather than only working for scalars. It still works for scalars, though.

```
from cyipopt import minimize_ipopt
def f(x):
    return x ** 2
minimize_ipopt(f, x0=1)
```
returning
```
     fun: 0.0
    info: {'x': array([0.]), 'g': array([], dtype=float64), 'obj_val': 0.0, 'mult_g': array([], dtype=float64), 'mult_x_L': array([0.]), 'mult_x_U': array([0.]), 'status': 0, 'status_msg': b'Algorithm terminated successfully at a locally optimal point, satisfying the convergence tolerances (can be specified by options).'}
 message: b'Algorithm terminated successfully at a locally optimal point, satisfying the convergence tolerances (can be specified by options).'
    nfev: 7
     nit: 1
    njev: 3
  status: 0
 success: True
       x: array([0.])
```

**Comment by mdhaber** [comment]
> Are the copy paste lines from SciPy necessary? Can you not import from SciPy instead?

Yeah they needed slight modification. Pretty much all of those changes are in the commit [TST: minimize_ipopt: make new tests pass](https://github.com/mechmotum/cyipopt/pull/207/commits/270af6aa0635068ef234f318cd9d3c16d31def40).

> We could include the SciPy license in the test directory for those two SciPy

Good call. The SciPy license is very permissive (New BSD) though. I can note which tests are copied from SciPy and include this license in the source file. Would you like me to do that?

**Comment by mdhaber** [comment]
> Yes. My general approach has always been to support the SciPy version that shipped with the most recent Ubuntu LTS. It looks like SciPy 1.8.0 was with 22.04 LTS.

Great. I think we can make that a separate PR, though.

File: cyipopt/scipy_interface.py [hunk 21]
```diff
@@ -604,21 +606,54 @@ def minimize_ipopt(fun,
...
     if not np.issubdtype(x0.dtype, np.number):
         raise ValueError('`x0` must be a numeric array.')
 
     # `method` does not need input validation. If `method is not None`, it is
     # passed to `scipy.optimize.minimize`, which raises a readable error if
     # the value isn't recognized.
 
-    # TODO: add input validation for `bounds` when adding
-    #  support for instances of new-style constraints (e.g. `Bounds`)
+    # Handle bounds that are either sequences (of sequences) or instances of
```
Comment (mdhaber) [inline cyipopt/scipy_interface.py:622]: I need to add to the documentation that `Bounds`, `LinearConstraints`, and `NonlinearConstraints` are now supported.

File: cyipopt/scipy_interface.py [hunk 22]
```diff
@@ -159,11 +159,13 @@ def wrapped_fun(x):
             con_hessian = con.get('hess', None)
             con_kwargs = con.get('kwargs', {})
             if con_jac is None:
+                con_jac = lambda x0, *args, **kwargs: optimize.approx_fprime(
+                    x0, con_fun, eps, *args, **kwargs)
```
Comment (mdhaber) [inline cyipopt/scipy_interface.py:163]:
```suggestion
```
This looks like it snuck in with a merge or something.

**Comment by mdhaber** [comment]
@moorepants I think I've replied to your comments. 

It sounds like I need to:
- Address these things I found (https://github.com/mechmotum/cyipopt/pull/207#pullrequestreview-1629657662)
- Add the SciPy license in the test directory. Shall I compbine `test_scipy_ipopt_from_scipy.py` and `test_scipy_ipopt_trust_constr.py` and include the license just in that file? I kept the files separate for now to facilitate review according to the suggestions in https://github.com/mechmotum/cyipopt/pull/207#issuecomment-1566406440, but if you're done, I think the files could be combined.

**Comment by moorepants** [comment]
> Add the SciPy license in the test directory. Shall I compbine test_scipy_ipopt_from_scipy.py and test_scipy_ipopt_trust_constr.py and include the license just in that file?

Yes, this sounds good. Then we can exclude that single file in the `MANIFEST.in` so it doesn't ship in the source tarball.

**Comment by mdhaber** [comment]
Done. Does GitHub save artifacts from the doc build? I'd like to see how those links rendered.

**Comment by moorepants** [comment]
> Done. Does GitHub save artifacts from the doc build? I'd like to see how those links rendered.

No, you have to build the docs yourself locally.

File: cyipopt/scipy_interface.py (hunk 1 above)
Comment (moorepants) [reply cyipopt/scipy_interface.py:538]: Ok. I consider this backwards incompatible, regardless of what the docs say it should do, but we can make the change with no deprecation.

**Comment by moorepants** [comment]
I'm good with this and can merge. I'll wait to see if you want to check the docs render.

Thanks for the addition! I'll work on a new release of cyipopt next.

**Comment by mdhaber** [comment]
The documentation changes here look fine.
image

The documentation could probably use another pass before the next release, but maybe let's make that a separate PR since it's unrelated.

**Comment by moorepants** [comment]
Sounds good. Merging.
